pkg/cri/resource-manager/resource-manager.go

```diff
@@ -53,188 +53,290 @@ type resmgr struct {
-		return resmgrError("failed to start controller: %v", err)
+	if err := m.startRequestProcessing(); err != nil {
+		return err
 	}
 
 	if err := m.relay.Start(); err != nil {
-		return resmgrError("failed to start relay: %v", err)
+		return resmgrError("failed to start CRI relay: %v", err)
 	}
 
-	if m.conf != nil {
-		m.cache.SetConfig(m.conf)
+	if opt.ForceConfig == "" {
+		if err := m.configServer.Start(opt.ConfigSocket); err != nil {
+			return resmgrError("failed to start configuration server: %v", err)
+		}
+
+		if m.conf != nil {
+			m.cache.SetConfig(m.conf)
+			m.conf = nil
+		}
 	}
 
-	m.Info("up and running...")
+	m.Info("up and running")
 
 	return nil
 }
 
 // Stop stops the resource manager.
 func (m *resmgr) Stop() {
-	m.relay.Client().Close()
-	m.relay.Server().Stop()
+	m.Info("shutting down...")
+
+	m.Lock()
+	defer m.Unlock()
+
+	if opt.ForceConfig == "" {
+		m.configServer.Stop()
+	}
+
+	m.relay.Stop()
 }
 
-// Set new resource manager configuration.
+// SetConfig pushes new configuration to the resource manager.
 func (m *resmgr) SetConfig(conf *config.RawConfig) error {
-	m.Info("received new configuration")
+	m.Info("applying new configuration...")
 
 	m.Lock()
 	defer m.Unlock()
 
 	// TODO: save current configuration and roll back if some controllers fail to start
 
 	if err := pkgcfg.SetConfig(conf.Data); err != nil {
-		m.Error("failed to update configuration: %v", err)
+		m.Error("new configuration was rejected: %v", err)
 		return resmgrError("configuration rejected: %v", err)
 	}
 
 	if err := m.control.StartStopControllers(m.cache, m.relay.Client()); err != nil {
-		m.Error("failed to activate configuration: %v", err)
+		m.Error("failed to activate new configuration: %v", err)
 		return resmgrError("failed to fully activate configuration: %v", err)
 	}
 
 	m.cache.SetConfig(conf)
-	m.Info("configuration successfully updated")
+	m.Info("successfully switched to new configuration")
+
+	return nil
+}
+
+// setupCache creates a cache and reloads its last saved state if found.
+func (m *resmgr) setupCache() error {
+	var err error
+
+	options := cache.Options{CacheDir: opt.RelayDir}
+	if m.cache, err = cache.NewCache(options); err != nil {
+		return resmgrError("failed to create cache: %v", err)
+	}
+
+	return nil
+
+}
+
+// setupConfigAgent sets up the connection to the configuration agent.
+func (m *resmgr) setupConfigAgent() error {
+	var err error
+
+	if m.agent, err = agent.NewAgentInterface(opt.AgentSocket); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// setupConfigServer sets up our configuration server for agent notifications.
+func (m *resmgr) setupConfigServer() error {
+	var err error
+
+	if opt.ForceConfig != "" {
+		m.Warn("using forced configuration %s, not starting configuration server",
+			opt.ForceConfig)
+	} else {
+		if m.configServer, err = config.NewConfigServer(m.SetConfig); err != nil {
+			return resmgrError("failed to create configuration notification server: %v", err)
+		}
+	}
```

Comment:
Related to my previous comment and thinking aloud: maybe we could remove the check here and create the `configServer` instance in any case. Then, just not start it in `resmgr.Start()`?

Reply:
Yes, maybe we should do just that. Actually I have a related/similar patch in the kubernetes-v1.17.2 PR for the agent. When we force the config, we should disable the agent interface as well. There I had the same dilemma: how to administer it. Change it to pointer and test for nil, etc. I ended up adding support for explicit disabling to the agent interface and it errors when called if it is disabled. 

If we create the configServer without starting here, then after the other patch lands these would be symmetric.

Reply:
Changed the code accordingly.